igor-web/src/test/groovy/com/netflix/spinnaker/igor/wercker/WerckerBuildMonitorSpec.groovy

```diff
@@ -0,0 +1,251 @@
+            new Run(id:"a",    startedAt: new Date(now-11), finishedAt: new Date(now-11)),
+            new Run(id:"init", startedAt: new Date(now-12), finishedAt: new Date(now-10)),
+        ]
+        monitor = monitor(buildMasters)
+        monitor.worker = scheduler.createWorker()
+        mockService.getRunsSince(_) >> [pipeline: runs1]
+        cache.getBuildNumber(_, _, _) >> 1
+
+        when:
+        monitor.onApplicationEvent(Mock(RemoteStatusChangedEvent))
+        scheduler.advanceTimeBy(1L, TimeUnit.SECONDS.MILLISECONDS)
+
+        then: 'initial poll'
+        1 * buildMasters.filteredMap(BuildServiceProvider.WERCKER) >> [MASTER: mockService]
+        1 * buildMasters.map >> [MASTER: mockService]
+        1 * echoService.postEvent(_)
+
+        cleanup:
+        monitor.stop()
+    }
+
+    void 'select latest one form multiple completed runs'() {
+        given:
+        cache.getJobNames(MASTER) >> ['pipeline']
+        BuildMasters buildMasters = Mock(BuildMasters)
+        buildMasters.map >> [MASTER: mockService]
+        monitor = monitor(buildMasters)
+        monitor.worker = scheduler.createWorker()
+        mockService.getRunsSince(_) >> [:]
+        cache.getBuildNumber(_, _, _) >> 1
+
+        when:
+        monitor.onApplicationEvent(Mock(RemoteStatusChangedEvent))
+        scheduler.advanceTimeBy(1L, TimeUnit.SECONDS.MILLISECONDS)
+
+        then: 'initial poll'
+        1 * buildMasters.filteredMap(BuildServiceProvider.WERCKER) >> [MASTER: mockService]
+        1 * buildMasters.map >> [MASTER: mockService]
+        0 * echoService.postEvent(_)
+
+        when:
+        scheduler.advanceTimeBy(998L, TimeUnit.SECONDS.MILLISECONDS)
+
+        then:
+        0 * buildMasters.map >> [MASTER: mockService]
+        0 * mockService.getRunsSince(_) >> [:]
+
+        when: 'poll at 1 second'
+        long now = System.currentTimeMillis();
+        List<Run> runs1 = [
+            new Run(id:"b",    startedAt: new Date(now-10)),
+            new Run(id:"a",    startedAt: new Date(now-11), finishedAt: new Date(now-11)),
+            new Run(id:"init", startedAt: new Date(now-12), finishedAt: new Date(now-10)),
+        ]
+        cache.getLastPollCycleTimestamp(_, _) >> (now - 1000)
+        mockService.getRunsSince(_) >> [pipeline: runs1]
+        cache.getBuildNumber(_, _, _) >> 1
+        scheduler.advanceTimeBy(2L, TimeUnit.SECONDS.MILLISECONDS)
+
+        then:
+        1 * buildMasters.filteredMap(BuildServiceProvider.WERCKER) >> [MASTER: mockService]
+        1 * buildMasters.map >> [MASTER: mockService]
+        1 * echoService.postEvent(_)
+
+        cleanup:
+        monitor.stop()
+    }
+
+    void 'get runs of multiple pipelines'() {
+        setup:
+        BuildMasters buildMasters = Mock(BuildMasters)
+        buildMasters.map >> [MASTER: werckerService]
+        monitor = monitor(buildMasters)
+        monitor.worker = scheduler.createWorker()
+        cache.getBuildNumber(_, _, _) >> 1
```

Comment:
This, and the next line, might be more maintainable with a wildcard `cache.getBuildNumber(*_)` so that we don't need to add an underscore if an argument is ever added to `getBuildNumber`

Reply:
fixed